cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -154,21 +156,39 @@ func NewClusterStateRegistry(cloudProvider cloudprovider.CloudProvider, config C
 	}
 }
 
-// RegisterScaleUp registers scale up.
-func (csr *ClusterStateRegistry) RegisterScaleUp(request *ScaleUpRequest) {
+// RegisterOrUpdateScaleUp registers scale-up for give node group or changes requested node increase
+// count.
+// If increaseDelta is positive then number of new nodes requested is increased; Time and expectedAddTime
+// are reset.
+// If increaseDelete is negative the number of new nodes requested is decreased; Time and expectedAddTime are
+// left intact.
+func (csr *ClusterStateRegistry) RegisterOrUpdateScaleUp(nodeGroup cloudprovider.NodeGroup, increaseDelta int, currentTime time.Time) {
 	csr.Lock()
 	defer csr.Unlock()
 
-	oldScaleUpRequest, found := csr.scaleUpRequests[request.NodeGroup.Id()]
+	scaleUpRequest, found := csr.scaleUpRequests[nodeGroup.Id()]
 	if !found {
-		csr.scaleUpRequests[request.NodeGroup.Id()] = request
-		return
+		scaleUpRequest = &ScaleUpRequest{
+			NodeGroup:       nodeGroup,
+			Time:            currentTime,
+			ExpectedAddTime: currentTime.Add(csr.config.MaxNodeProvisionTime),
+		}
+		csr.scaleUpRequests[nodeGroup.Id()] = scaleUpRequest
 	}
 
 	// update the old request
-	oldScaleUpRequest.Time = request.Time
-	oldScaleUpRequest.ExpectedAddTime = request.ExpectedAddTime
-	oldScaleUpRequest.Increase += request.Increase
+	if scaleUpRequest.Increase+increaseDelta <= 0 {
+		// increase <= 0 means that there is no scale-up intent really
```

Comment:
Do we expect to get negative numbers here? Should we just silently ignore it if we do?

Reply:
I think it can if someone manually increase MIG size and then the nodes cannot be provisioned because of stockout/quota/whatever.